### ISSUE TITLE
TypeError: Invalid array length argument

### ISSUE THREAD

**Issue by ptarjan** [jhiesey/stream-http#29]
In this chunk of code:

https://github.com/jhiesey/stream-http/blob/7d1dfd2ab16280a223a8778eef257fa3594f306e/lib/response.js#L137

```
  case 'arraybuffer':
    if (xhr.readyState !== rStates.DONE)
      break
    response = xhr.response
    self.push(new Buffer(new Uint8Array(response)))
    break
```

It assumes the response is a valid integer. I'm seeing cases with some of my users where the `Uint8Array(response)` is throwing 

```
TypeError: Invalid array length argument
```

Should you check the response first or do you intend to throw an exception in this case and it is the caller's responsibility to catch exceptions like this?

**Comment by jhiesey** [comment]
In this code, `response` is actually supposed to be an `ArrayBuffer`. This certainly shouldn't be throwing exceptions.

Could you give me a reproducible example that has this issue?

**Comment by ptarjan** [comment]
Sadly I don't have a consistent repro, it is just bubbling up in my logs every few days. I don't know how your library is set-up but the other side of a connection can send whatever they want right? If so, the client should guard against random crap on the wire.

**Comment by ansongoldade** [comment]
I get this intermittently as well and can't reproduce. I've only seen it with Safari so far but that's anecdotal. Might I suggest wrapping the following line with a try catch and emitting an error in the catch scenario so that "error" code can take the proper recovery steps? 

https://github.com/jhiesey/stream-http/blob/7d1dfd2ab16280a223a8778eef257fa3594f306e/lib/request.js#L206

In the current state when the referenced error occurs it is not emitted on the request or the response. If you're in a polling situation, this is the end of the road and you're forced to respond to the error at some other level in the application and that doesn't feel like the right place.

**Comment by feross** [comment]
@jhiesey One simple solution might be to change this line (https://github.com/jhiesey/stream-http/blob/7d1dfd2ab16280a223a8778eef257fa3594f306e/lib/response.js#L141) to just pass the array buffer directly into the `Buffer` constructor, since that's supported now, without doing a copy :)

That way, if the data randomly happens to be a number, it'll return a zeroed out array instead of throwing an exception. Not the correct behavior, but probably better than a random exception in potentially async code?

**Comment by jhiesey** [comment]
My guess is that `response` is null, which is allowed by the spec but I'm not checking for. I'll put in a check for that plus a try/catch just in case.

**Comment by jhiesey** [comment]
I suspect this is another duplicate of #20 since I wasn't checking that the request hadn't failed. If so, it should be fixed as of v2.0.3.

If it's still present just reopen and I'll take another look.
